core/domain/prod_validation_jobs_one_off.py

```diff
@@ -5038,6 +5039,133 @@ def _get_custom_validation_functions(cls):
         ]
 
 
+class PlaythroughModelValidator(BaseModelValidator):
+    """Class for validating PlaythroughModel."""
+
+    PLAYTHROUGH_PROJECT_RELEASE_DATETIME = datetime.datetime(2018, 9, 1)
+
+    @classmethod
+    def _get_external_id_relationships(cls, item):
+        exp_id = item.exp_id
+        exp_version = item.exp_version
+        exp_issues_id = (
+            stats_models.ExplorationIssuesModel.get_entity_id(
+                exp_id, exp_version)
+        )
+
+        return {
+            'exp_ids': (
+                exp_models.ExplorationModel, [item.exp_id]),
+            'exp_issues': (
+                stats_models.ExplorationIssuesModel, [exp_issues_id])
+        }
+
+    @classmethod
+    def _get_model_id_regex(cls, unused_item):
+        return r'^[A-Za-z0-9-_]{1,%s}\.[A-Za-z0-9-_]{1,%s}$' % (
+            base_models.ID_LENGTH, base_models.ID_LENGTH)
+
+    @classmethod
+    def _get_model_domain_object_instance(cls, item):
+        return stats_services.get_playthrough_from_model(item)
+
+    @classmethod
+    def _validate_exploration_id_in_whitelist(cls, item):
+        """Validate the exploration id in playthrough model is in
+        the whitelist.
+
+        Args:
+            item: ndb.Model. PlaythroughModel to validate.
+        """
+        whitelisted_exp_ids_for_playthroughs = (
+            config_domain.WHITELISTED_EXPLORATION_IDS_FOR_PLAYTHROUGHS.value)
+
+        if item.exp_id not in whitelisted_exp_ids_for_playthroughs:
+            cls.errors['exploration id check'].append(
+                'Entity id %s: recorded in exploration_id:%s which '
+                'has not been curated for recording.' % (
+                    item.id, item.exp_id)
+            )
+
+    @classmethod
+    def _validate_reference(cls, item):
+        """Validate the playthrough reference relations.
+
+        Args:
+            item: ndb.Model. PlaythroughModel to validate.
+        """
+        exp_issues_tuples = cls.external_instance_details['exp_issues']
+        for _, _, exp_issues in exp_issues_tuples:
+            # The case for missing ExplorationIssues external model is
+            # ignored here since errors for missing email external model
+            # are already checked and stored in
+            # _validate_external_id_relationships function.
+            exp_id = item.exp_id
+            exp_version = item.exp_version
+
+            issues = []
+            for issue in exp_issues.unresolved_issues:
+                issue_type = issue['issue_type']
+                if (
+                        item.id in issue['playthrough_ids']
+                        and issue_type == item.issue_type):
```

Comment:
Let's make this more strict, exactly one 'playthrough_id' should exist in any of the unresolved issues.

Reply:
I'm not quite understand this.
Do you mean `assert len(issue['playthough_ids']) == 1` or that there should not be any duplicate playthrough id in `issue['playthrough_ids']` in each unresolved issue?

Reply:
The latter, there should not be any duplicate playthrough IDs in any of the unresolved issues. We can make a stronger check in the ExplorationIssuesModel audit to make sure eacy `playthrough_id` appears exactly once, but let's do the weaker check here anyway.

Reply:
For such cases, should we yield errors like 'for playthrough id=x, its corresponding ExplorationIssuesModel (id=y), has unresolved issues with duplicate playthrough ids, which are [z1, z1, ...]'?
This error message seems not so relevant to the PlaythroughModel being audited, I feel we should keep this type of errors in the audit job for ExplorationIssuesModel.

And, if there are duplicated `item.id` (the PlaythroughModel instance that is under validation), we are already yielding errors like `it is referenced multiple time in an issue`. If there are duplicated ids of other instances of PlaythroughModel, then we will have the same messages during their own validation procedure.